runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java

```diff
@@ -37,13 +37,19 @@
 /**
  * Holds user state in memory. Only one key is active at a time due to the GroupReduceFunction being
  * called once per key. Needs to be reset via {@code resetForNewKey()} before processing a new key.
+ *
+ * <p>In case of any failures, this factory must be discarded. Otherwise, the contained state cache
+ * token would be reused which would corrupt the state cache.
  */
 public class InMemoryBagUserStateFactory<K, V, W extends BoundedWindow>
     implements StateRequestHandlers.BagUserStateHandlerFactory<K, V, W> {
 
+  private final ByteString cacheToken;
+
   private List<InMemorySingleKeyBagState> handlers;
 
   public InMemoryBagUserStateFactory() {
+    cacheToken = ByteString.copyFrom(UUID.randomUUID().toString().getBytes(Charsets.UTF_8));
```

Comment:
I think we'll want to move the cache token generation up higher to the `ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter`. This comment added for this class should be moved to the ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter

We should also test that it only adds a single cache token to the request.

Reply:
Could you add a reason for moving this up the stack? State handlers should implement their own cache token handling.

Reply:
I'll have another look later.

Reply:
`ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter#getCacheTokens` is going to add N copies of the cache token to the `ProcessBundleRequest`

For user state, it makes sense to have a caching handler that does the delegation to other handlers. This caching handler is the one that should be responsible for supplying the single cache token.

Reply:
I see, makes sense. Let me update the PR.